### 프로젝트(1) load_digits:손글씨를 분류해 봅시다

#### (1) 필요한 모듈 import하기

In [3]:
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

#### (2) 데이터 준비

In [4]:
digits = load_digits()

#### (3) 데이터 이해하기

In [12]:
digits_data = digits.data
digits_label = digits.target
digits.target
digits_data.shape

(1797, 64)

#### (4) train, test 데이터 분리

In [13]:
X_train, X_test, y_train, y_test = train_test_split(digits_data, 
                                                    digits_label, 
                                                    test_size=0.2, 
                                                    random_state=7)


#### (5) 다양한 모델로 학습시켜보기

##### 1. Decision Tree 사용해 보기

In [14]:
from sklearn.tree import DecisionTreeClassifier

              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       0.81      0.81      0.81        42
           2       0.79      0.82      0.80        40
           3       0.79      0.91      0.85        34
           4       0.83      0.95      0.89        37
           5       0.90      0.96      0.93        28
           6       0.84      0.93      0.88        28
           7       0.96      0.82      0.89        33
           8       0.88      0.65      0.75        43
           9       0.78      0.78      0.78        32

    accuracy                           0.86       360
   macro avg       0.86      0.86      0.86       360
weighted avg       0.86      0.86      0.85       360



##### 2. Random Forest 사용해 보기

In [16]:
from sklearn.ensemble import RandomForestClassifier

In [18]:
random_forest = RandomForestClassifier(random_state=32)
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       0.93      1.00      0.97        42
           2       1.00      1.00      1.00        40
           3       1.00      1.00      1.00        34
           4       0.93      1.00      0.96        37
           5       0.90      0.96      0.93        28
           6       1.00      0.96      0.98        28
           7       0.94      0.97      0.96        33
           8       1.00      0.84      0.91        43
           9       0.94      0.94      0.94        32

    accuracy                           0.96       360
   macro avg       0.96      0.96      0.96       360
weighted avg       0.97      0.96      0.96       360



##### 3. SVM 사용해 보기

In [19]:
from sklearn import svm

In [20]:
svm_model = svm.SVC()

svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        43
           1       0.95      1.00      0.98        42
           2       1.00      1.00      1.00        40
           3       1.00      1.00      1.00        34
           4       1.00      1.00      1.00        37
           5       0.93      1.00      0.97        28
           6       1.00      1.00      1.00        28
           7       1.00      1.00      1.00        33
           8       1.00      0.93      0.96        43
           9       1.00      0.97      0.98        32

    accuracy                           0.99       360
   macro avg       0.99      0.99      0.99       360
weighted avg       0.99      0.99      0.99       360



##### 4. SGD Classifier 사용해 보기

In [22]:
from sklearn.linear_model import SGDClassifier

In [23]:
sgd_model = SGDClassifier()

sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        43
           1       0.88      0.88      0.88        42
           2       0.95      0.97      0.96        40
           3       0.85      1.00      0.92        34
           4       0.95      0.95      0.95        37
           5       0.97      1.00      0.98        28
           6       0.93      0.93      0.93        28
           7       1.00      0.88      0.94        33
           8       0.95      0.86      0.90        43
           9       0.94      0.94      0.94        32

    accuracy                           0.94       360
   macro avg       0.94      0.94      0.94       360
weighted avg       0.94      0.94      0.94       360



##### 5. Logistic Regression 사용해 보기

In [24]:
from sklearn.linear_model import LogisticRegression

In [25]:
logistic_model = LogisticRegression()

logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        43
           1       0.95      0.95      0.95        42
           2       0.98      1.00      0.99        40
           3       0.94      0.97      0.96        34
           4       0.97      1.00      0.99        37
           5       0.82      0.96      0.89        28
           6       1.00      0.96      0.98        28
           7       0.97      0.97      0.97        33
           8       0.92      0.81      0.86        43
           9       0.97      0.91      0.94        32

    accuracy                           0.95       360
   macro avg       0.95      0.95      0.95       360
weighted avg       0.95      0.95      0.95       360



/home/aiffel/Downloads/ls/envs/aiffel/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:765: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


#### (6) 모델을 평가해 보기

모델 성능을 평가하는 지표로 f1-score accuracy를 선택했다. 왜냐하면, f1-score는 precison과 recall의 조화평균값이기 때문이다. Decision Tree 모델을 사용하였을 때가 0.86으로 가장 낮았고, 그 다음으로 SGD Classifier 모델을 사용하였을 때가 0.94로 높아진 결과를 얻을 수 있었다. 그 다음으로는 Logistic Regression 모델을 사용하였을 때가 0.95로 높아졌고, Random Forest 모델을 사용하였을 때가 0.96으로 두번째로 높은 값을 보여주었다. 마지막으로, SVM 모델을 사용하였을 때가 0.99로 제일 높은 Accuracy를 나타내었다. 

### 프로젝트(2) load_wine:와인을 분류해 봅시다

#### (1) 필요한 모듈 import하기

In [26]:
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

#### (2) 데이터 준비

In [28]:
wine = load_wine()

#### (3) 데이터 이해하기

In [49]:
wine_data = wine.data
wine_label = wine.target
wine_data.shape

(178, 13)

In [31]:
list(wine.target_names)

['class_0', 'class_1', 'class_2']

#### (4) train, test 데이터 분리

In [34]:
X_train, X_test, y_train, y_test = train_test_split(wine_data, 
                                                    wine_label, 
                                                    test_size=0.2, 
                                                    random_state=7)

#### (5) 다양한 모델로 학습시켜보기

##### 1. Decision Tree 사용해 보기

In [35]:
from sklearn.tree import DecisionTreeClassifier

In [36]:
decision_tree = DecisionTreeClassifier(random_state=32)
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00         7
           1       0.89      1.00      0.94        17
           2       1.00      0.83      0.91        12

    accuracy                           0.94        36
   macro avg       0.96      0.94      0.95        36
weighted avg       0.95      0.94      0.94        36



##### 2. Random Forest 사용해 보기

In [38]:
from sklearn.ensemble import RandomForestClassifier

In [39]:
random_forest = RandomForestClassifier(random_state=32)
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00         7
           1       1.00      1.00      1.00        17
           2       1.00      1.00      1.00        12

    accuracy                           1.00        36
   macro avg       1.00      1.00      1.00        36
weighted avg       1.00      1.00      1.00        36



##### 3. SVM 사용해 보기

In [40]:
from sklearn import svm

In [41]:
svm_model = svm.SVC()

svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.86      0.86      0.86         7
           1       0.58      0.88      0.70        17
           2       0.33      0.08      0.13        12

    accuracy                           0.61        36
   macro avg       0.59      0.61      0.56        36
weighted avg       0.55      0.61      0.54        36



##### 4. SGD Classifier 사용해 보기

In [42]:
from sklearn.linear_model import SGDClassifier

In [43]:
sgd_model = SGDClassifier()

sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.43      0.60         7
           1       0.00      0.00      0.00        17
           2       0.36      1.00      0.53        12

    accuracy                           0.42        36
   macro avg       0.45      0.48      0.38        36
weighted avg       0.32      0.42      0.29        36



/home/aiffel/Downloads/ls/envs/aiffel/lib/python3.7/site-packages/sklearn/metrics/_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/home/aiffel/Downloads/ls/envs/aiffel/lib/python3.7/site-packages/sklearn/metrics/_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/home/aiffel/Downloads/ls/envs/aiffel/lib/python3.7/site-packages/sklearn/metrics/_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start,

##### 5. Logistic Regression 사용해 보기

In [44]:
from sklearn.linear_model import LogisticRegression

In [45]:
logistic_model = LogisticRegression()

logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.86      0.92         7
           1       0.94      1.00      0.97        17
           2       1.00      1.00      1.00        12

    accuracy                           0.97        36
   macro avg       0.98      0.95      0.96        36
weighted avg       0.97      0.97      0.97        36



/home/aiffel/Downloads/ls/envs/aiffel/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:765: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


#### (6) 모델을 평가해 보기

모델 성능을 평가하는 지표로 f1-score accuracy를 선택했다. 왜냐하면, f1-score는 precison과 recall의 조화평균값이기 때문이다. SGD Classifier 모델을 사용하였을 때가 0.42 가장 낮았고, 그 다음으로 SVM 모델을 사용하였을 때가 0.61로 높아진 결과를 얻을 수 있었다. 그 다음으로는 Decision Tree 모델을 사용하였을 때가 0.94로 높아졌고, Logistic Regression 모델을 사용하였을 때가 0.97으로 두번째로 높은 값을 보여주었다. 마지막으로, Random Forest 모델을 사용하였을 때가 1.00로 제일 높은 Accuracy를 나타내었다. 

### 프로젝트(3)load_breast_cancer:유방암 여부를 진단해 봅시다

#### (1) 필요한 모듈 import하기

In [46]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

#### (2) 데이터 준비

In [47]:
breast_cancer = load_breast_cancer()

#### (3) 데이터 이해하기

In [51]:
breast_cancer_data = breast_cancer.data
breast_cancer_label = breast_cancer.target
breast_cancer_data.shape

(569, 30)

In [52]:
list(breast_cancer.target_names)

['malignant', 'benign']

#### (4) train, test 데이터 분리

In [53]:
X_train, X_test, y_train, y_test = train_test_split(breast_cancer_data, 
                                                    breast_cancer_label, 
                                                    test_size=0.2, 
                                                    random_state=7)

#### (5) 다양한 모델로 학습시켜보기

##### 1. Decision Tree 사용해 보기

In [54]:
from sklearn.tree import DecisionTreeClassifier

In [55]:
decision_tree = DecisionTreeClassifier(random_state=32)
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.92      0.82      0.87        40
           1       0.91      0.96      0.93        74

    accuracy                           0.91       114
   macro avg       0.91      0.89      0.90       114
weighted avg       0.91      0.91      0.91       114



##### 2. Random Forest 사용해 보기

In [56]:
from sklearn.ensemble import RandomForestClassifier

In [57]:
random_forest = RandomForestClassifier(random_state=32)
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        40
           1       1.00      1.00      1.00        74

    accuracy                           1.00       114
   macro avg       1.00      1.00      1.00       114
weighted avg       1.00      1.00      1.00       114



##### 3.SVM 사용해 보기

In [58]:
from sklearn import svm

In [59]:
svm_model = svm.SVC()

svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.72      0.84        40
           1       0.87      1.00      0.93        74

    accuracy                           0.90       114
   macro avg       0.94      0.86      0.89       114
weighted avg       0.92      0.90      0.90       114



##### 4.SGD Classifier 사용해 보기

In [60]:
from sklearn.linear_model import SGDClassifier

In [61]:
sgd_model = SGDClassifier()

sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.85      0.82      0.84        40
           1       0.91      0.92      0.91        74

    accuracy                           0.89       114
   macro avg       0.88      0.87      0.87       114
weighted avg       0.89      0.89      0.89       114



##### 5.Logistic Regression 사용해 보기

In [62]:
from sklearn.linear_model import LogisticRegression

In [63]:
logistic_model = LogisticRegression()

logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.85      0.92        40
           1       0.93      1.00      0.96        74

    accuracy                           0.95       114
   macro avg       0.96      0.93      0.94       114
weighted avg       0.95      0.95      0.95       114



/home/aiffel/Downloads/ls/envs/aiffel/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:765: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


#### (6) 모델을 평가해 보기

모델 성능을 평가하는 지표로 f1-score accuracy를 선택했다. 왜냐하면, f1-score는 precison과 recall의 조화평균값이기 때문이다. SGD Classifier 모델을 사용하였을 때가 0.89로 가장 낮았고, 그 다음으로 SVM 모델을 사용하였을 때가 0.90로 높아진 결과를 얻을 수 있었다. 그 다음으로는 Decision Tree 모델을 사용하였을 때가 0.91로 높아졌고, Logistic Regression 모델을 사용하였을 때가 0.95으로 두번째로 높은 값을 보여주었다. 마지막으로, Random Forest 모델을 사용하였을 때가 1.00로 제일 높은 Accuracy를 나타내었다. 

### 느낀점(회고)
손글씨 이미지 데이터를 분류하는 것과 와인 분류하기/유방암여부 진단하기가 결과가 다르게 나오고, 와인 분류하기와 유방암 여부 진단하기 결과가 비슷하게 나온 것을 확인할 수 있었다. SVM 모델은 손글씨 이미지 데이터 분류하는데 ACCURACY가 높게 나온 반면, 와인 분류하기와 유방암 분류하는데에서는 낮은 ACCURACY를 보였다. 데이터의 개수가 제일 작은 와인 분류하기에서는 제일 낮은 ACCURACY를 보여주고 있다. 대체로 Random Forest 모델은 세가지 경우에 높은 ACCURACY를 보여주었다. 다양한 MACHINE LEARNING 모델들을 테스트해볼 수 있는 좋은 기회였다.